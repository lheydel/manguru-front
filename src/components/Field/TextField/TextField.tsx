import React from 'react';
import MuiTextField, { TextFieldProps as MuiTextFieldProps } from '@material-ui/core/TextField';
import { FormikProps } from 'formik';

export interface TextFieldProps {
  name: string;
  errors: FormikProps<any>['errors'];
  touched: FormikProps<any>['touched'];
  values: FormikProps<any>['values'];
  handleBlur: FormikProps<any>['handleBlur'];
  handleChange: FormikProps<any>['handleChange'];
}

/**
 * Combine Material-UI TextField design with Formik functionnalities
 */
export const TextField: React.FC<TextFieldProps & MuiTextFieldProps> = ({
  name,
  errors,
  touched,
  values,
  handleBlur,
  handleChange,
  ...props
}) => (
  <MuiTextField
    name={name}
    error={!!errors[name] && !!touched[name]}
    helperText={touched[name] && errors[name]}
    value={values[name]}
    onChange={handleChange}
    onBlur={handleBlur}
    {...props}
  />
);
