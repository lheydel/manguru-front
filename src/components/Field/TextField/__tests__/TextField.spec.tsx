import React from 'react';
import { TextFieldProps, TextField } from '../TextField';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('TextField component', () => {
  const props: (touched?: boolean, error?: boolean) => TextFieldProps = (touched, error) => {
    const name = 'blblbl';
    return {
      errors: { [name]: touched && error ? 'err' : '' },
      name: name,
      touched: { [name]: touched },
      values: {},
      handleBlur: jest.fn(),
      handleChange: jest.fn(),
    };
  };

  it('should render itself and children', () => {
    const wrapper = shallow(<TextField {...props()} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render text field with no error', () => {
    const wrapperTF = shallow(<TextField {...props(true, false)} />);
    const wrapperFT = shallow(<TextField {...props(false, true)} />);
    expect(toJson(wrapperTF)).toMatchSnapshot();
    expect(toJson(wrapperFT)).toMatchSnapshot();
  });

  it('should render text field with error', () => {
    const wrapper = shallow(<TextField {...props(true, true)} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should trigger handleBlur', () => {
    const fieldProps = props();
    const wrapper = shallow(<TextField {...fieldProps} />);
    wrapper.simulate('blur');
    expect(fieldProps.handleBlur).toHaveBeenCalled();
  });

  it('should trigger handleChange', () => {
    const fieldProps = props();
    const wrapper = shallow(<TextField {...fieldProps} />);
    wrapper.simulate('change');
    expect(fieldProps.handleChange).toHaveBeenCalled();
  });
});
