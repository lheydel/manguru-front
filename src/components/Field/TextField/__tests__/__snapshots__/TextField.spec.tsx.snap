// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField component should render itself and children 1`] = `
<WithStyles(ForwardRef(TextField))
  error={false}
  name="blblbl"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
/>
`;

exports[`TextField component should render text field with error 1`] = `
<WithStyles(ForwardRef(TextField))
  error={true}
  helperText="err"
  name="blblbl"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
/>
`;

exports[`TextField component should render text field with no error 1`] = `
<WithStyles(ForwardRef(TextField))
  error={false}
  helperText=""
  name="blblbl"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
/>
`;

exports[`TextField component should render text field with no error 2`] = `
<WithStyles(ForwardRef(TextField))
  error={false}
  helperText={false}
  name="blblbl"
  onBlur={[MockFunction]}
  onChange={[MockFunction]}
/>
`;
